src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SnippetTaglet.java

```diff
@@ -0,0 +1,354 @@
+import com.sun.source.doctree.DocTree;
+import com.sun.source.doctree.SnippetTree;
+import com.sun.source.doctree.TextTree;
+import jdk.javadoc.doclet.Taglet;
+import jdk.javadoc.internal.doclets.toolkit.Content;
+import jdk.javadoc.internal.doclets.toolkit.DocletElement;
+import jdk.javadoc.internal.doclets.toolkit.taglets.snippet.Action;
+import jdk.javadoc.internal.doclets.toolkit.taglets.snippet.ParseException;
+import jdk.javadoc.internal.doclets.toolkit.taglets.snippet.Parser;
+import jdk.javadoc.internal.doclets.toolkit.taglets.snippet.StyledText;
+import jdk.javadoc.internal.doclets.toolkit.util.Utils;
+
+import javax.lang.model.element.Element;
+import javax.lang.model.element.PackageElement;
+import javax.tools.Diagnostic;
+import javax.tools.DocumentationTool.Location;
+import javax.tools.FileObject;
+import javax.tools.JavaFileManager;
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * A taglet that represents the {@code @snippet} tag.
+ *
+ *  <p><b>This is NOT part of any supported API.
+ *  If you write code that depends on this, you do so at your own risk.
+ *  This code and its internal interfaces are subject to change or
+ *  deletion without notice.</b>
+ */
+public class SnippetTaglet extends BaseTaglet {
+
+    public SnippetTaglet() {
+        super(DocTree.Kind.SNIPPET, true, EnumSet.allOf(Taglet.Location.class));
+    }
+
+    /*
+     * A snippet can specify content by value or by reference.
+     *
+     * To specify content by value, a snippet uses its body; the body of a snippet is the content.
+     * To specify content by reference, a snippet uses either the "class" or "file" attribute;
+     * the value of that attribute refers to the content.
+     *
+     * A snippet can specify the "region" attribute. That attribute refines the location of the content.
+     * The value of that attribute must match one of the named regions in the snippets content.
+     */
+    @Override
+    public Content getInlineTagOutput(Element holder, DocTree tag, TagletWriter writer) {
+        SnippetTree snippetTag = (SnippetTree) tag;
+
+        Map<String, AttributeTree> attributes = new HashMap<>();
+
+        // organize attributes in a map, performing basic checks along the way
+        for (DocTree d : snippetTag.getAttributes()) {
+            if (!(d instanceof AttributeTree a)) {
+                continue; // this might be an ErroneousTree
+            }
+            if (attributes.putIfAbsent(a.getName().toString(), a) == null) {
+                continue;
+            }
+            // two like-named attributes found; although we report on the most
+            // recently encountered of two, the iteration order might differ
+            // from the source order
+            error(writer, holder, a, "doclet.tag.attribute.repeated", a.getName().toString());
```

Comment:
Can we use a `LinkedHashMap` or similar to preserve encounter order?

Reply:
If you look closely, you will see that the ordering characteristics of the `Map` implementation used to collect attributes is immaterial. My comment was about the implied (but unspecified) order of attributes in the list returned from the `SnippetTree.getAttributes()` method; see the third paragraph in the description of https://bugs.openjdk.java.net/browse/JDK-8266826

I clarified the comment in 26de34083a3.